lib/Scheduling/CPSATSchedulers.cpp

```diff
@@ -0,0 +1,138 @@
+                                        Operation *lastOp) {
+  Operation *containingOp = prob.getContainingOp();
+  if (!prob.hasOperation(lastOp))
+    return containingOp->emitError("problem does not include last operation");
+
+  CpModelBuilder cpModel;
+  auto &tasks = prob.getOperations();
+  auto &resources = prob.getOperatorTypes();
+
+  DenseMap<Operation *, IntVar> taskStarts;
+  DenseMap<Operation *, IntVar> taskEnds;
+  DenseMap<Problem::OperatorType, llvm::SmallVector<IntervalVar, 4>>
+      resourcesToTaskIntervals;
+
+  // first get horizon
+  unsigned horizon = 0;
+  for (auto *task : tasks) {
+    unsigned duration = *prob.getLatency(*prob.getLinkedOperatorType(task));
+    horizon += duration;
+  }
+
+  // build vars
+  for (auto item : llvm::enumerate(tasks)) {
+    auto i = item.index();
+    auto *task = item.value();
+    IntVar startVar = cpModel.NewIntVar(Domain(0, horizon))
+                          .WithName((Twine("start_of_task_") + Twine(i)).str());
+    IntVar endVar = cpModel.NewIntVar(Domain(0, horizon))
+                        .WithName((Twine("end_of_task_") + Twine(i)).str());
+    taskStarts[task] = startVar;
+    taskEnds[task] = endVar;
+    auto resource = prob.getLinkedOperatorType(task);
+    if (resource.hasValue()) {
+      unsigned duration = *prob.getLatency(*resource);
+      IntervalVar taskInterval =
+          cpModel.NewIntervalVar(startVar, duration, endVar)
+              .WithName((Twine("task_interval_") + Twine(i)).str());
+
+      if (!resourcesToTaskIntervals.count(resource.value()))
+        resourcesToTaskIntervals[resource.value()] = {};
+      resourcesToTaskIntervals[resource.value()].emplace_back(taskInterval);
+    }
+  }
+
+  IntVar makeSpan = cpModel.NewIntVar(Domain(0, horizon)).WithName("makespan");
+  for (Operation *task : tasks) {
+    cpModel.AddLessOrEqual(taskEnds[task], makeSpan);
+    for (auto dep : prob.getDependences(task)) {
+      Operation *src = dep.getSource();
+      Operation *dst = dep.getDestination();
+      if (src == dst)
+        return containingOp->emitError() << "dependence cycle detected";
+      cpModel.AddLessOrEqual(taskEnds[src], taskStarts[dst]);
+    }
+  }
+
+  for (auto resourceToTaskIntervals : resourcesToTaskIntervals) {
+    // https://or.stackexchange.com/a/3363
+    const Problem::OperatorType &resource = resourceToTaskIntervals.getFirst();
+    auto capacity = prob.getLimit(resource);
+    if (!capacity)
+      continue;
+
+    const llvm::SmallVector<IntervalVar, 4> &taskIntervals =
+        resourceToTaskIntervals.getSecond();
+
+    CumulativeConstraint cumu = cpModel.AddCumulative(capacity.value());
+    for (auto &item : llvm::enumerate(taskIntervals)) {
+      auto i = item.index();
+      auto taskInterval = item.value();
+      IntVar demandVar = cpModel.NewIntVar(Domain(1)).WithName(
+          (Twine("demand_") + Twine(i) + Twine("_") + Twine(resource.str()))
```

Comment:
Micro-Nit: `Twine('_') + Twine(resource.strref())`

Reply:
:D does the `.str()` incur a copy or something like that as opposed to the `strref()`? Probably.

Reply:
`.str()` builds an `std::string`, whereas `.strref()` returns some kind of reference to the data in the unique'd `StringAttr` (I think 😇 ). Again, a micro-nit :wink:

Reply:
>Again, a micro-nit

The best kind :)